pkg/agent/controller/networkpolicy/filestore_test.go

```diff
@@ -0,0 +1,190 @@
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/protobuf"
+	"k8s.io/apimachinery/pkg/types"
+
+	"antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+)
+
+const (
+	testDirNetworkPolicy  = "/var/run/antrea-test/file-store/network-policies"
+	testDirAppliedToGroup = "/var/run/antrea-test/file-store/applied-to-groups"
+	testDirAddressGroup   = "/var/run/antrea-test/file-store/address-groups"
+)
+
+// Set it to MemMapFs as the file system may be not writable.
+// Change it to NewOsFs to evaluate performance when writing to disk.
+var newFS = afero.NewMemMapFs
+
+func newFakeFileStore(tb testing.TB, path string) *fileStore {
+	serializer := protobuf.NewSerializer(scheme, scheme)
+	codec := codecs.CodecForVersions(serializer, serializer, v1beta2.SchemeGroupVersion, v1beta2.SchemeGroupVersion)
+	// Create a new FS for every fileStore in case of interaction between tests.
+	s, err := newFileStore(newFS(), path, codec)
+	assert.NoError(tb, err)
+	return s
+}
+
+func TestFileStore(t *testing.T) {
+	policy1 := newNetworkPolicy("policy1", "uid1", []string{"addressGroup1"}, nil, []string{"appliedToGroup1"}, nil)
+	policy2 := newNetworkPolicy("policy2", "uid2", []string{"addressGroup2"}, nil, []string{"appliedToGroup2"}, nil)
+	policy3 := newNetworkPolicy("policy3", "uid3", []string{"addressGroup3"}, nil, []string{"appliedToGroup3"}, nil)
+	updatedPolicy2 := policy2.DeepCopy()
+	updatedPolicy2.AppliedToGroups = []string{"foo"}
+
+	tests := []struct {
+		name            string
+		ops             func(*fileStore)
+		expectedObjects []runtime.Object
+	}{
+		{
+			name: "add",
+			ops: func(store *fileStore) {
+				store.save(policy1)
+				store.save(policy2)
+				store.save(policy3)
+			},
+			expectedObjects: []runtime.Object{policy1, policy2, policy3},
+		},
+		{
+			name: "update",
+			ops: func(store *fileStore) {
+				store.save(policy1)
+				store.save(policy2)
+				store.save(updatedPolicy2)
+			},
+			expectedObjects: []runtime.Object{policy1, updatedPolicy2},
+		},
+		{
+			name: "delete",
+			ops: func(store *fileStore) {
+				store.save(policy1)
+				store.save(policy2)
+				store.delete(policy2)
+			},
+			expectedObjects: []runtime.Object{policy1},
+		},
+		{
+			name: "replace",
+			ops: func(store *fileStore) {
+				store.save(policy1)
+				store.save(policy2)
+				store.replaceAll([]runtime.Object{updatedPolicy2, policy3})
+			},
+			expectedObjects: []runtime.Object{updatedPolicy2, policy3},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := newFakeFileStore(t, testDirNetworkPolicy)
+			tt.ops(s)
+			gotObjects, err := s.loadAll()
+			require.NoError(t, err)
+			assert.Equal(t, tt.expectedObjects, gotObjects)
+		})
+	}
+}
+
+func BenchmarkFileStoreAddNetworkPolicy(b *testing.B) {
+	policy1 := newNetworkPolicy("policy1", "uid1", []string{"addressGroup1"}, nil, []string{"appliedToGroup1"}, nil)
```

Comment:
I think we can get more realistic mem usage if we use actually uid length string here, and also worth noting the benchmark result is based on single rule per policy

Reply:
UID doesn't make much difference as it's the number of members in a group plays the most important role. I have added UID to each object and the numbers are almost the same.

The number of rules in a policy wouldn't cause much difference too as long as we are not talking about hundreds level and above. It may increase 2-3us per policy write with 10 more rules in each policy. It won't change the number of MB in size. 

The test isn't aimed to cover various use cases but to give a rough impression about the level of the extra cost introduced by the storage. I think 16us to write a policy justifies it's not something we need to care about.